### PR TITLE
Wordpress & WooCommerce Implementation

### PR DESCRIPTION
**Pull request by rondoor124**
This is the initial version of the WooCommerce element.
[#3135](https://github.com/XRFoundation/XREngine/issues/3135)

### CODE REVIEW

**Comment by speigg** [comment]
@rondoor124 is this still a draft?

**Comment by rondoor124** [comment]
Hi @speigg, this branch is ready for the QA, I will host the WordPress site somewhere and push the world file soon,

File: woocommerce.world [hunk 1]
```diff
@@ -0,0 +1 @@
...:"model","extend":{"src":"https://xrengine.000webhostapp.com/wp-content/uploads/2021/11/razer_laptop.glb","textureOverride":"","castShadow":false,"receiveShadow":false,"loop-animation":{"activeClipIndex":-1,"hasAvatarAnimations":false}}}},{"name":"interact","props":{"interactable":true,"interactionType":"infoBox","interactionText":"ThinkPad X1 Carbon Gen 9 Intel (14\")","interactionThemeIndex":0,"interactionName":"ThinkPad X1 Carbon Gen 9 Intel (14\")","interactionDescription":"With the Intel® Evo™ platform, the ThinkPad X1 Carbon Gen 9 laptop delivers a powerhouse combination of performance, responsiveness, battery life, and stunning visuals. With up to 11th Gen Intel® Core™ i7 vPro® processors, you can count on an exceptional experience, anywhere.\n","interactionImages":["https://xrengine.000webhostapp.com/wp-content/uploads/2021/11/image-3-300x200.jpg"],"interactionVideos":["https://xrengine.000webhostapp.com/wp-content/uploads/2021/11/Sample.mp4"],"interactionUrls":["https://xrengine.000webhostapp.com/product/12"],"interactionModels":["https://xrengine.000webhostapp.com/wp-content/uploads/2021/11/razer_laptop.glb"]}}],"parent":"2266BED7-6CC4-48A6-95DD-9BCD3CF9EAFC","index":7}}}
```
Comment (rondoor124) [inline woocommerce.world:1]: This is the Woocommerce world file. Please be careful about the CORS when fetching the assets.

**Comment by rondoor124** [comment]
Hi, @speigg This PR is ready for review. Thanks.

**Comment by speigg** [comment]
We need to find a solution for this 
Screen Shot 2021-11-21 at 6 12 49 PM

**Review by speigg — APPROVED** [review]

**Comment by speigg** [comment]
@rondoor124  Just a couple merge conflicts w/ dev. Can you fix that, and then I'll merge this in? Thanks!

**Review by HexaField — CHANGES_REQUESTED** [review]
We can simplify the shopify and woocommerce scene loaded cases by removing them entirely and changing the json data to use the image, video, gltf-model components directly, serialising them on an as-needed basis in the node editors. We have optimizations being made in [this PR](https://github.com/XRFoundation/XREngine/pull/4248) which will make these entirely composable (including editor nodes) so we will have no duplicated code in the codebase anywhere. It is fine for now, but lets try and keep the scene loader as simple as possible.

We can also remove the cors wrapper in loadGLTFModel where the conflict exists.

File: packages/engine/src/scene/classes/Video.ts [hunk 2]
```diff
@@ -91,7 +92,7 @@ export default class Video extends AudioSource {
         // })
         // resolve()
       } else {
-        this.el.src = this._src
+        this.el.src = corsAnywhereUrl(this._src)
```
Comment (HexaField) [inline packages/engine/src/scene/classes/Video.ts:95]: no longer necessary

File: packages/engine/src/assets/functions/loadTexture.ts [hunk 3]
```diff
@@ -1,9 +1,12 @@
 import { TextureLoader } from 'three'
 import { isClient } from '../../common/functions/isClient'
+import { corsAnywhereUrl } from '@xrengine/client-core/src/util/cors'
 
 export default function loadTexture(src, textureLoader: any = new TextureLoader()) {
   return new Promise((resolve, reject) => {
     if (!isClient) resolve(null)
-    textureLoader.load(src, resolve, null, (error) => reject(new Error(`Error loading texture "${src}"`)))
+    textureLoader.load(corsAnywhereUrl(src), resolve, null, (error) =>
```
Comment (HexaField) [inline packages/engine/src/assets/functions/loadTexture.ts:8]: no longer necessary

File: packages/engine/src/assets/functions/LoadGLTF.ts [hunk 4]
```diff
@@ -30,7 +31,7 @@ export function disposeDracoLoaderWorkers(): void {
 export async function LoadGLTF(url: string): Promise<LoadGLTFResultInterface> {
   return await new Promise<LoadGLTFResultInterface>((resolve, reject) => {
     getLoader().load(
-      url,
+      corsAnywhereUrl(url),
```
Comment (HexaField) [inline packages/engine/src/assets/functions/LoadGLTF.ts:34]: no longer necessary

File: packages/client-core/src/util/cors.tsx [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import { Config } from '@xrengine/common/src/config'
+
+let corsAddress = ''
+if (process.env.APP_ENV == 'development') {
+  corsAddress = `https://${process.env.VITE_SERVER_HOST}:${process.env.VITE_CORS_SERVER_PORT}`
+} else {
+  corsAddress = `${Config.publicRuntimeConfig.apiServer}/cors-proxy/`
+}
+
+export const corsAnywhereUrl = (url) => {
```
Comment (HexaField) [inline packages/client-core/src/util/cors.tsx:10]: this file can be removed

File: .env.local.default [hunk 6]
```diff
@@ -28,6 +28,7 @@ VITE_APP_HOST=localhost
 VITE_APP_PORT=3000
 VITE_SERVER_HOST=localhost
 VITE_SERVER_PORT=3030
+VITE_CORS_SERVER_PORT=3035
```
Comment (HexaField) [inline .env.local.default:31]: no longer necessary

File: packages/editor/src/nodes/WooCommerceNode.ts [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+
+    const media = this.wooCommerceProductItems[index].media
+    this.extendType = media.extendType
+
+    if (media.extendType == 'model') {
+      this.extendNode = new ModelNode()
+      this.extendNode.initialScale = 'fit'
+    } else if (media.extendType == 'video') {
+      this.extendNode = new VideoNode()
+      this.extendNode.src = corsAnywhereUrl(media.url)
```
Comment (HexaField) [inline packages/editor/src/nodes/WooCommerceNode.ts:312]: no longer necessary

File: packages/editor/src/nodes/WooCommerceNode.ts [hunk 8]
```diff
@@ -0,0 +1,507 @@
...
+    } else if (media.extendType == 'image') {
+      this.extendNode = new ImageNode()
+    }
+
+    await this.extendNode.load(corsAnywhereUrl(media.url))
```
Comment (HexaField) [inline packages/editor/src/nodes/WooCommerceNode.ts:317]: no longer necessary

File: packages/engine/src/scene/functions/loadGLTFModel.ts [hunk 9]
```diff
@@ -198,7 +199,7 @@ export const loadGLTFModel = (
     new Promise<void>((resolve, reject) => {
       AssetLoader.load(
         {
-          url: component.data.src,
+          url: corsAnywhereUrl(component.data.src),
```
Comment (HexaField) [inline packages/engine/src/scene/functions/loadGLTFModel.ts:202]: no longer necessary

File: packages/ops/xrengine/templates/api-server-ingress.yaml [hunk 10]
```diff
@@ -1,6 +1,7 @@
 {{- if and .Values.api.enabled (ne (.Values.api.ingress.disabled | default "") "true") -}}
 {{- $fullName := include "xrengine.api.fullname" . -}}
 {{- $svcPort := .Values.api.service.port -}}
+{{- $corsSvcPort := .Values.api.service.corsServerPort -}}
```
Comment (HexaField) [inline packages/ops/xrengine/templates/api-server-ingress.yaml:4]: please ensure this is merged correctly

File: packages/projects/default-project/wooCommerce.scene.json [hunk 11]
```diff
@@ -0,0 +1,584 @@
+{
+  "version": 4,
+  "root": "2266BED7-6CC4-48A6-95DD-9BCD3CF9EAFC",
+  "entities": {
```
Comment (HexaField) [inline packages/projects/default-project/wooCommerce.scene.json:4]: lets not worry about a new scene for now, we are already showing off shopify intergration

**Review by HexaField — CHANGES_REQUESTED** [review]

**Review by HexaField — APPROVED** [review]

**Pull request merged**